webhook/resources/volume/validator.go

```diff
@@ -0,0 +1,149 @@
+	return admission.Resource{
+		Name:       "volumes",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Volume{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (v *volumeValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	volume := newObj.(*longhorn.Volume)
+
+	if !util.ValidateName(volume.Name) {
+		return werror.NewInvalidError(fmt.Sprintf("invalid name %v", volume.Name), "")
+	}
+
+	if err := types.ValidateDataLocality(volume.Spec.DataLocality); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateAccessMode(volume.Spec.AccessMode); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateReplicaCount(volume.Spec.NumberOfReplicas); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if err := types.ValidateReplicaAutoBalance(volume.Spec.ReplicaAutoBalance); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	if volume.Spec.BackingImage != "" {
+		if _, err := v.ds.GetBackingImage(volume.Spec.BackingImage); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if volume.Spec.EngineImage == "" {
+		return werror.NewInvalidError("BUG: Invalid empty Setting.EngineImage", "")
+	}
+
+	if !volume.Spec.Standby {
+		if volume.Spec.Frontend != longhorn.VolumeFrontendBlockDev && volume.Spec.Frontend != longhorn.VolumeFrontendISCSI {
+			return werror.NewInvalidError(fmt.Sprintf("invalid volume frontend specified: %v", volume.Spec.Frontend), "")
+		}
+	}
+
+	if volume.Spec.Migratable && volume.Spec.AccessMode != longhorn.AccessModeReadWriteMany {
+		return werror.NewInvalidError("migratable volumes are only supported in ReadWriteMany (rwx) access mode", "")
+	}
+
+	// Check engine version before disable revision counter
+	if volume.Spec.RevisionCounterDisabled {
+		if ok, err := v.canDisableRevisionCounter(volume.Spec.EngineImage); !ok {
+			err := errors.Wrapf(err, "can not create volume with current engine image that doesn't support disable revision counter")
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	currentNodeID, err := util.GetRequiredEnv(types.EnvNodeName)
+	if err != nil {
+		return werror.NewInvalidError("BUG: fail to detect the node name", "")
+	}
+
+	m := manager.NewVolumeManager(currentNodeID, v.ds)
+	if volume.Spec.DataSource != "" {
+		if err := m.VerifyDataSourceForVolumeCreation(volume.Spec.DataSource, volume.Spec.Size); err != nil {
+			return werror.NewInvalidError(err.Error(), "")
+		}
+	}
+
+	if err := datastore.CheckVolume(volume); err != nil {
+		return werror.NewInvalidError(err.Error(), "")
+	}
+
+	return nil
+}
+
+func (v *volumeValidator) Update(request *admission.Request, oldObj runtime.Object, newObj runtime.Object) error {
```

Comment:
Notice that this function has both the old obj and the new obj, we can move more validations from the update APIs to this validator. e.g., once there is a change for field `Spec.NumberOfReplicas`, we need to apply these checks https://github.com/longhorn/longhorn-manager/blob/c04eac1c804597c78d5962c1758e8ed80d2abb37/manager/volume.go#L1016-L1024

And this logic can be applied to other validators. If this would take time, you can open one more PR for it.

Reply:
> And this logic can be applied to other validators. If this would take time, you can open one more PR for it.

I will process this part in a separate PR.